### PR TITLE
fix(providers): surface cause-specific provider failure diagnostics (#9001)

### LINKED ISSUE #9001: [Bug]: Provider turn failures bury cause-specific diagnostics under a generic retry envelope
### Affected component
provider

### Severity
S2 - degraded behavior

### Current behavior
Terminal provider failures surface through the same `All model_providers/models failed` envelope followed by raw retry traces.

This affects materially different conditions, including:

- LM Studio or Ollama not running
- Missing API credentials
- Rejected or expired credentials
- Unknown models
- Rate limiting
- Timeouts
- Remote connectivity failures

The reliable-provider layer already classifies some underlying errors, but the primary user-facing message remains generic. Missing-key wording is also not consistently recognized as an authentication failure.

### Expected behavior
The primary message should identify the actionable cause and provider, for example:

- `LM Studio is not reachable at http://localhost:1234. Start its local server or update the endpoint.`
- `No API key is configured for OpenAI. Add credentials for the selected provider.`
- `The configured credentials for Anthropic were rejected.`
- `The configured model was not found by Ollama.`

Detailed retry and fallback information should remain available in logs without dominating Zerocode's chat output. Existing retry and fallback behavior should remain unchanged.

### Steps to reproduce
1. Select a local provider while its server is stopped and send a message.
2. Select a cloud provider without an API key and send a message.
3. Repeat with rejected credentials or an unknown model.
4. Observe that each failure leads with the same generic envelope.

### Impact
Users must interpret low-level retry traces to discover routine configuration or local-process problems, and may be directed toward irrelevant VPN or firewall troubleshooting.

### Related work
#6889 restored reqwest source-chain details. #7515 added the current structured provider diagnostics. This issue concerns cause-specific user presentation.

### ZeroClaw version
Current master

### Operating system
macOS

### Regression?
Unknown

### Pre-flight checks
- [x] Confirmed in current master source and recent dogfood testing.
- [x] No secrets or personal data included.

### PR DESCRIPTION
**Pull request by wm0018**
## Summary

- **Base branch:** `master`
- **What changed and why:** Terminal provider failures all surfaced through the generic `All model_providers/models failed` envelope followed by raw retry traces, so routine problems (a local server not running, a missing/rejected API key, an unknown model, a timeout) were indistinguishable without reading low-level output. The reliable layer already classified the underlying error into a `kind` but never promoted it to the user-facing message. The provider edge crate now emits a structured, locale-neutral `ProviderFailureLead` (cause kind + provider + sanitized endpoint) as the `bail!` lead line; the runtime renders that lead through its Fluent catalog into a concise, actionable sentence at the user-facing boundary, and the full per-attempt envelope stays in structured logs only.
- **What changed and why:** Auth is split into `auth_missing` / `auth_rejected` via a display-only `classify_auth_kind`, with a combined honest `auth` fallback. Rejection rides on the portable HTTP 401/403 contract; missing is best-effort (some providers, e.g. local/OpenAI-compatible, never emit a pre-flight "no key" error). The public `is_auth_error` predicate is intentionally left unchanged so the orchestrator's cache-eviction control flow is not widened.
- **What changed and why:** Missing-credential wording (including Anthropic's real `credentials not set`) and stringified connection failures (`connection refused`, etc., which do not always survive as a typed `reqwest::Error`) are now recognized as their true cause instead of the generic fallback.
- **What changed and why:** Connect/DNS guidance distinguishes local vs remote endpoints via `is_localhost_endpoint`, so a cloud API outage is never told to "start its local server".
- **Scope boundary:** Retry and fallback behavior is unchanged (`is_non_retryable` / retry counts untouched); `is_auth_error` semantics are unchanged. The gateway `ws.rs` `error_code` substring classifier is intentionally left for a follow-up (out of scope here).
- **Blast radius:** `crates/zeroclaw-providers` (structured lead + classifier) and `crates/zeroclaw-runtime` (Fluent rendering at `dispatch.rs` + 5 locale catalogs). ZeroCode renders the runtime-supplied `TurnComplete.content` verbatim, so it gets the concise lead for free; nothing parses the old envelope for control flow.
- **Linked issue(s):** Closes #9001
- **Labels:** `bug`, `provider`, `runtime`, `provider:reliable`, `priority:p2`, `risk:medium`, `size:M`

### How you can test (when useful)

- **Reviewer testing requested?** Yes
- **Interface(s) exercised:** `cli` (`zeroclaw agent -a <alias> -m ...`); same error path feeds `tui` (zerocode) and `web` (gateway).
- **Setup / preconditions:** An isolated config dir (`ZEROCLAW_CONFIG_DIR`) with one agent and a failing provider.
- **Steps to run:**
  - Missing key: configure `[providers.models.anthropic.<alias>]` with a model but no key, then `zeroclaw agent -a <alias> -m hello`.
  - Unreachable local: configure a `custom` provider with `uri = "http://127.0.0.1:1/v1"` and any placeholder key, then send a message.
  - Unreachable remote: configure a `custom` provider with a cloud `uri` that refuses/DNS-fails, then send a message.
- **Expected on this branch (after):**
  - Missing key → `No API key is configured for anthropic. Add credentials for the selected provider.`
  - Unreachable local → `custom is not reachable at http://127.0.0.1:1/v1/chat/completions. Start its local server or update the endpoint.`
  - Unreachable remote → `custom is not reachable at https://api.example.com/.... Check the network, VPN, or firewall, or update the endpoint.`
  - The full `All model_providers/models failed. Attempts: ...` envelope is **no longer** in `TurnComplete.content`; it remains in structured logs only.
- **Prior behavior on `master` (before):** Both lead with `All model_providers/models failed. Attempts:` (unreachable) or `The request to anthropic failed. See the logs for details.` (missing key), i.e. no cause-specific lead.

### How I tested

- **CI checks relied on and why they cover this change:** The workspace Rust build/test/clippy/fmt legs of `ci.yml` cover both changed crates (`zeroclaw-providers`, `zeroclaw-runtime`). The comment hygiene gate enforces no issue refs leak into code comments.
- **Known CI coverage gap, if any:** None for the changed surface.
- **Commands run and tail output:**

```sh
cargo test -p zeroclaw-providers --lib
# test result: ok. 1187 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

cargo test -p zeroclaw-runtime --lib i18n::tests::provider_failure
# 3 passed; 0 failed (boundary regression: no envelope in rendered lead; local vs remote guidance)

cargo test -p zeroclaw-runtime --lib rpc::dispatch
# test result: ok. 165 passed; 0 failed

cargo clippy -p zeroclaw-providers -p zeroclaw-runtime --all-targets -- -D warnings
# clean

cargo fmt -p zeroclaw-providers -p zeroclaw-runtime -- --check
# clean
```

- **Beyond CI, what did you manually verify?** Boundary unit tests assert the rendered lead contains no `All model_providers/models failed` / `Attempts:` and that a remote endpoint is not told to start a local server (while a local endpoint is). The 12 pre-existing locale-sensitive runtime failures on `master` (zh-CN environment, English assertions in `safety_net`/`agent`/`daemon`/`doctor`/`service`/`quickstart`) are unrelated to this change — verified by reproducing one on clean `master`.
- **If any command was intentionally skipped, why:** N/A.

## Security & Privacy Impact (required)

- New permissions, capabilities, or file system access scope? No
- New external network calls? No
- Secrets / tokens / credentials handling changed? No — messages name the provider and (for connect/timeout) the sanitized endpoint via the existing `sanitize_api_error` / `sanitized_url_endpoint` paths; no credential values are surfaced.
- PII, real identities, or personal data in diff, tests, fixtures, or docs? No
- Prompt injection or untrusted model-visible text introduced/changed? No

## Compatibility (required)

- Backward compatible? Yes — error text is human-facing only; no consumer parses the envelope for control flow. `is_auth_error` semantics are unchanged, so orchestrator cache-eviction behavior is unchanged.
- Config / env / CLI surface changed? No
- Rust/MSRV/toolchain floor changed? No

## Rollback (required for medium/high-risk PRs)

- **Fast rollback command/path:** `git revert <sha>`
- **Feature flags or config toggles:** None
- **Observable failure symptoms:** Provider-failure messages regress to the generic `All model_providers/models failed` lead in `TurnComplete.content`; grep logs for `error_kind` attrs to confirm classification still fires.

### CODE REVIEW

**Comment by wm0018** [comment]
### Live terminal logs (patched binary, isolated config)

Captured with a `debug` build of this branch against a throwaway `ZEROCLAW_CONFIG_DIR`, so no real credentials are involved. Exit code is `1` in both cases. The lead line is the new actionable summary; the full per-attempt envelope (with `kind=`) is retained below it, and retry counts are unchanged.

**Scenario A — local server not running (`kind=connect`)**

```text
$ zeroclaw agent -a repro -m "hello"
Error: custom is not reachable at http://127.0.0.1:1/v1/chat/completions. Start its local server or update the endpoint.

All model_providers/models failed. Attempts:
model_provider=custom model=test-model attempt 1/3: retryable; error=error sending request for url (http://127.0.0.1:1/v1/chat/completions): client error (Connect): tcp connect error: Connection refused (os error 111); kind=connect; phase=connect; hint=could not open provider connection; check network, VPN, or firewall; endpoint=http://127.0.0.1:1/v1/chat/completions
model_provider=custom model=test-model attempt 2/3: retryable; error=... (same) ...
model_provider=custom model=test-model attempt 3/3: retryable; error=... (same) ...
```

**Scenario B — missing API key (`kind=auth_missing`)**

```text
$ zeroclaw agent -a repro -m "hello"
Error: No API key is configured for anthropic. Add credentials for the selected provider.

All model_providers/models failed. Attempts:
model_provider=anthropic model=claude-sonnet-4-6 attempt 1/3: retryable; error=Anthropic credentials not set. Set ANTHROPIC_API_KEY or ANTHROPIC_OAUTH_TOKEN (setup-token).; kind=auth_missing; phase=http_response; hint=check provider credentials
model_provider=anthropic model=claude-sonnet-4-6 attempt 2/3: retryable; error=... (same) ...
model_provider=anthropic model=claude-sonnet-4-6 attempt 3/3: retryable; error=... (same) ...
```

`auth_rejected` (a real 401/403) requires a live remote call and is covered by unit tests rather than a live run here.

**Comment by wm0018** [comment]
### Live terminal logs — remaining scenarios (patched binary, isolated config)

Same setup as above: a `debug` build of this branch against a throwaway `ZEROCLAW_CONFIG_DIR`, no real credentials (exit code `1`). HTTP-status cases use a tiny local mock server; the timeout case uses a mock that stalls past `timeout_secs`. The lead line is the new actionable summary; the full per-attempt envelope (with `kind=`) is retained below it, and retry counts are unchanged.

**Scenario C — rejected credentials (`kind=auth_rejected`, mock `401`)**

```text
$ zeroclaw agent -a repro -m "hello"
Error: The configured credentials for custom were rejected.

All model_providers/models failed. Attempts:
model_provider=custom model=test-model attempt 1/3: non_retryable; error=Custom API error (401 Unauthorized): {"error": {"message": "Invalid API key provided", "type": "invalid_request_error", "code": "invalid_api_key"}}; kind=auth_rejected; phase=http_response; hint=check provider credentials
```

**Scenario D — unknown model (`kind=model_not_found`, mock `404`)**

```text
$ zeroclaw agent -a repro -m "hello"
Error: The configured model was not found by custom. Check the model id for the selected provider.

All model_providers/models failed. Attempts:
model_provider=custom model=test-model attempt 1/3: non_retryable; error=Custom API error (404 Not Found): {"error": {"message": "The model `ghost-model` does not exist", "type": "invalid_request_error", "code": "model_not_found"}}; kind=model_not_found; phase=http_response; hint=check the configured model id for this provider
```

**Scenario E — rate limiting (`kind=rate_limited`, mock `429`)**

```text
$ zeroclaw agent -a repro -m "hello"
Error: custom rate-limited the request. Wait, change key/quota, or switch provider.

All model_providers/models failed. Attempts:
model_provider=custom model=test-model attempt 1/3: rate_limited; error=Custom API error (429 Too Many Requests): {"error": {"message": "Rate limit reached for requests", "type": "rate_limit_error", "code": "rate_limit_exceeded"}}; kind=rate_limited; phase=http_response; hint=wait, change key/quota, or switch provider
model_provider=custom model=test-model attempt 2/3: rate_limited; error=Custom API error (429 Too Many Requests): {"error": {"message": "Rate limit reached for requests", "type": "rate_limit_error", "code": "rate_limit_exceeded"}}; kind=rate_limited; phase=http_response; hint=wait, change key/quota, or switch provider
model_provider=custom model=test-model attempt 3/3: rate_limited; error=Custom API error (429 Too Many Requests): {"error": {"message": "Rate limit reached for requests", "type": "rate_limit_error", "code": "rate_limit_exceeded"}}; kind=rate_limited; phase=http_response; hint=wait, change key/quota, or switch provider
```

**Scenario F — timeout (`kind=timeout`, slow mock + `timeout_secs=2`)**

```text
$ zeroclaw agent -a repro -m "hello"
Error: custom did not respond in time at http://127.0.0.1:8095/v1/chat/completions. Retry, check the network, or switch provider.

All model_providers/models failed. Attempts:
model_provider=custom model=test-model attempt 1/3: retryable; error=error sending request for url (http://127.0.0.1:8095/v1/chat/completions): operation timed out; kind=timeout; phase=request; hint=provider request timed out; retry or switch provider; endpoint=http://127.0.0.1:8095/v1/chat/completions
model_provider=custom model=test-model attempt 2/3: retryable; error=error sending request for url (http://127.0.0.1:8095/v1/chat/completions): operation timed out; kind=timeout; phase=request; hint=provider request timed out; retry or switch provider; endpoint=http://127.0.0.1:8095/v1/chat/completions
model_provider=custom model=test-model attempt 3/3: retryable; error=error sending request for url (http://127.0.0.1:8095/v1/chat/completions): operation timed out; kind=timeout; phase=request; hint=provider request timed out; retry or switch provider; endpoint=http://127.0.0.1:8095/v1/chat/completions
```

DNS resolution failure shares the `connect` "not reachable" message in practice (reqwest reports DNS failures as connect errors) and the dedicated `dns` kind is covered by the unit test `provider_error_diagnostic_classifies_text_error_branches`.

File: crates/zeroclaw-providers/src/reliable.rs [hunk 1]
```diff
@@ -1109,7 +1266,10 @@ impl ModelProvider for ReliableModelProvider {
         }
 
         anyhow::bail!(
-            "All model_providers/models failed. Attempts:\n{}",
+            "{}\n\nAll model_providers/models failed. Attempts:\n{}",
```
Comment (IftekharUddin) [inline crates/zeroclaw-providers/src/reliable.rs]: ### 🔴 Blocking — Keep retry traces out of ZeroCode chat

This still returns the full per-attempt envelope as part of the user-facing error. On the current direct-turn path, `crates/zeroclaw-runtime/src/rpc/dispatch.rs` puts `format!("turn failed: {e}")` into a failed `TurnComplete`, and `apps/zerocode/src/chat.rs` renders that `content` verbatim. The author’s manual output confirms the result is a concise first line followed by the same multi-line retry dump. Please keep the full attempts in structured logs while returning a separate concise diagnostic to the surface, and add an RPC/ZeroCode boundary test that rejects `All model_providers/models failed. Attempts:` in the displayed failed-turn content.

File: crates/zeroclaw-providers/src/reliable.rs [hunk 2]
```diff
@@ -350,6 +414,77 @@ struct ProviderErrorDiagnostic {
...
 }
 
+impl ProviderErrorDiagnostic {
+    /// Build a single, actionable, provider-named lead sentence for end users
+    /// (#9001). The reliable layer already classifies the underlying error into
+    /// `kind`; this promotes that classification into a message a user can act
+    /// on, instead of leading with the generic "All model_providers/models
+    /// failed" envelope. Detailed retry/fallback traces stay untouched in the
+    /// logs and in the tail of the bail message.
+    fn user_facing_summary(&self, provider: &str) -> String {
```
Comment (IftekharUddin) [inline crates/zeroclaw-providers/src/reliable.rs]: ### 🔴 Blocking — Localize at the owning presentation boundary

This helper is explicitly user-facing, but every branch hard-codes English below the runtime and ZeroCode localization owners. `AGENTS.md` requires runtime user-facing text to use Fluent, and `docs/book/src/architecture/localization-catalog-lifecycle.md` documents ZeroCode’s independent catalog. A provider-layer string cannot select either surface’s locale. Please keep the diagnostic structured here and render it through the owning surface catalog(s), with catalog/fallback coverage for the new messages.

File: crates/zeroclaw-providers/src/reliable.rs [hunk 3]
```diff
@@ -350,6 +414,77 @@ struct ProviderErrorDiagnostic {
...
+                "No API key is configured for {provider}. Add credentials for the selected provider."
+            ),
+            "auth_rejected" => {
+                format!("The configured credentials for {provider} were rejected.")
+            }
+            "auth" => format!(
+                "The configured credentials for {provider} were rejected or missing. Add or check the API key for the selected provider."
+            ),
+            "connect" | "dns" => format!(
+                "{provider} is not reachable{at}. Start its local server or update the endpoint."
```
Comment (IftekharUddin) [inline crates/zeroclaw-providers/src/reliable.rs]: ### 🔴 Blocking — Do not prescribe a local server for remote failures

`connect` and `dns` are not evidence that this is a local provider. The same branches cover a cloud API, remote proxy, or DNS outage, so an OpenAI/Anthropic connectivity failure can now tell the user to “Start its local server.” #9001 explicitly includes remote connectivity failures. Please distinguish a known-local case or use guidance that is truthful for both local and remote endpoints, and add a remote-endpoint regression case.

File: crates/zeroclaw-providers/src/reliable.rs [hunk 4]
```diff
@@ -186,6 +186,17 @@ pub fn is_auth_error(err: &anyhow::Error) -> bool {
         "403 forbidden",
         "invalid api key",
         "incorrect api key",
+        // A missing/unconfigured key is an authentication failure too, not a
+        // generic provider error (#9001). Providers phrase this several ways,
+        // including "credentials not set" (Anthropic) as well as "api key".
+        "missing api key",
```
Comment (IftekharUddin) [inline crates/zeroclaw-providers/src/reliable.rs]: ### 🟡 Warning — This changes cache-eviction control flow

These new phrases do more than improve classification for display: `zeroclaw-channels/src/orchestrator/mod.rs` calls this public predicate to decide whether to evict the cached provider. Missing-key errors will now take that branch. That behavior may be intentional, but it contradicts the body’s “only presentation” / “nothing ... for control flow” claims. Please either use a separate display classifier or explicitly document and test the intended cache eviction, then update the public scope and compatibility claims.

**Review by IftekharUddin — CHANGES_REQUESTED** [review]
I reviewed 3c8c1e81ff50f38d50a44d1fc24c5eb4b37fbad3 against the linked issue, the direct-turn RPC/ZeroCode path, the localization contract, and current `master`. Unless a finding includes a **Follow-up path**, it needs to be corrected on this branch; opening an issue alone does not resolve it.

### 🔴 Blocking — Keep retry traces out of ZeroCode chat

The cause-specific lead is useful, but each final `bail!` still concatenates the complete per-attempt envelope. The RPC path puts `e.to_string()` into the failed `TurnComplete`, and ZeroCode renders that content verbatim as a system message. The author’s own terminal captures show the resulting multi-line retry dump. This therefore does not satisfy #9001’s requirement that detail remain in logs without dominating chat.

**Why this matters:** the affected user surface still exposes the low-level output this change is intended to replace.

**If ignored:** ZeroCode users will see a better first line followed by the same noisy traces, and `Closes #9001` will overstate what the PR delivers.

**Required change:** preserve the complete attempt data in structured logs, but give direct-turn surfaces a separate concise diagnostic. Add a boundary test proving a failed ZeroCode/RPC turn contains the actionable summary without the retry envelope.

### 🔴 Blocking — Localize at the owning presentation boundary

`user_facing_summary` introduces a new set of hard-coded English product messages inside `zeroclaw-providers`. The repository contract requires runtime user-facing text to use Fluent, while ZeroCode has its own independent Fluent catalog. Formatting these strings below both presentation layers makes them impossible for either loader to translate correctly.

**Why this matters:** provider diagnostics reach CLI, TUI, web, and channel surfaces with different localization owners.

**If ignored:** non-English users receive English-only failure guidance, and later wording changes have no catalog source of truth.

**Required change:** keep a structured cause/provider/endpoint diagnostic at the provider boundary, then render the user-facing sentence through each owning surface’s localization contract. Cover the new keys and the relevant fallback/loading behavior.

### 🔴 Blocking — Do not prescribe a local server for remote failures

The shared `connect | dns` summary always says “Start its local server,” but those kinds also cover cloud-provider DNS failures, refused proxy connections, and other remote connectivity failures explicitly listed in #9001.

**Why this matters:** this PR’s core contract is actionable, cause-specific guidance.

**If ignored:** a cloud-provider outage or bad DNS path sends the user toward an impossible local-process fix.

**Required change:** either distinguish cases where a local process is actually known or use guidance that remains true for both local and remote endpoints. Add a remote-endpoint connect/DNS case alongside the local-server case.

### 🔴 Blocking — Refresh the conflicted head

GitHub currently reports this branch as `CONFLICTING` / `DIRTY`. Current `master` substantially changed `reliable.rs` for model-pinned and streaming fallback reporting in #8684. I resolved the textual conflict in a synthetic merge; formatting passed and `cargo test -p zeroclaw-providers --lib` passed all 1,154 tests, but the contributor branch still needs to be refreshed and the final head revalidated before it can merge.

### 🟡 Warning — The change is not presentation-only

The added missing-credential phrases broaden the public `is_auth_error` predicate. `zeroclaw-channels` uses that predicate to evict the cached provider, so missing-key failures now alter cache/rebuild control flow. That may be desirable, but it contradicts the PR body’s “only presentation changes” and “nothing parses ... for control flow” claims. Please either separate the display classifier from the cache-refresh predicate or document and test the intended eviction behavior; then correct the scope, compatibility, CI-gap, and closing claims in the body.

### 🟢 What looks good — Classification evidence is focused

The missing-versus-rejected auth distinction, typed 401/403 precedence, reuse of endpoint sanitization, focused provider tests, and six-scenario manual evidence are all useful. Required CI is green on the submitted head, and the provider library remains green in the synthetic current-`master` merge.

**Comment by Audacity88** [comment]
@wm0018, the July 23 review remains unanswered. The branch still needs to keep retry traces out of user-facing chat, render diagnostics through the owning localization surfaces, use guidance that works for both local and remote connection failures, refresh the conflicting branch, and correct the auth-error control-flow and PR-body claims.

Marking this as `stale-candidate`. Please update the branch or reply with your plan by **August 12, 2026**. #9001 remains open and accepted; closing this branch would not reject the provider-diagnostics fix.

**Review by Audacity88 — CHANGES_REQUESTED** [review]
@wm0018, thanks for updating this. The current head fixes the earlier auth-cache control-flow regression and no longer gives local-server guidance for remote endpoints. Moving user-facing sentence selection into the Fluent layer is also the right direction.

I re-reviewed head `8ee300b25a0237fc037c90aa24ff1a1067bca8b7` against the prior requested changes, linked issue #9001, and the actual CLI, RPC/ZeroCode, channel, and web error paths. The RPC path is improved, but the concise localized projection is not yet the common behavior described by the PR, and the new marker encoding makes the unrepaired paths worse for users.

The remaining work is concentrated in the error boundary and its user-facing proof. Once these two blockers are repaired, we can validate the real CLI and RPC/ZeroCode behavior without reopening the already-resolved concerns.

### ✅ Resolved — Cache control flow and remote guidance

Missing credentials are now display-only classification and no longer widen auth rejection or cache eviction. Connect and DNS failures also distinguish local from remote endpoints, so cloud-provider failures no longer tell users to start a local server.

### 🟢 What looks good — Cause-specific diagnostics are moving to the right layer

The cause prioritization is focused, the missing-versus-rejected credential distinction is useful, and the runtime Fluent catalog is the correct owner for user-facing sentences. Those pieces are worth preserving while the transport between provider errors and presentation surfaces is repaired.

### 🔴 Blocking — Complete the user-facing provider-failure projection

The new renderer is only called from RPC dispatch. Direct `zeroclaw agent` execution propagates or prints the provider error unchanged, and immediate context-window exits bypass `ProviderFailureLead`, causing RPC to fall back to the original error text. Channel and WebSocket consumers also retain separate raw-error paths even though the PR body says the same repaired path feeds the web dashboard. These paths can therefore expose the encoded `provider-failure|...` marker and/or the retry envelope instead of the concise localized cause.

Please replace the delimiter-in-`anyhow` protocol with a typed provider-owned error or another fail-closed structured boundary, then ensure every surface claimed by this PR receives the concise projection and every terminal exit, including context-window failures, uses it. The current parser searches for a marker anywhere and trusts reconstructed values, so it is not a robust structured boundary. Add focused tests that drive the actual CLI and RPC/ZeroCode boundaries and prove their user-visible output contains neither the marker nor the retry envelope while diagnostic records retain the attempt details. Open PR #9557 is also defining a typed provider-error taxonomy, so please coordinate the ownership boundary rather than landing parallel classifiers.

Relevant paths: [`reliable.rs`](https://github.com/zeroclaw-labs/zeroclaw/blob/8ee300b25a0237fc037c90aa24ff1a1067bca8b7/crates/zeroclaw-providers/src/reliable.rs#L415), [`dispatch.rs`](https://github.com/zeroclaw-labs/zeroclaw/blob/8ee300b25a0237fc037c90aa24ff1a1067bca8b7/crates/zeroclaw-runtime/src/rpc/dispatch.rs#L1994), [`loop_.rs`](https://github.com/zeroclaw-labs/zeroclaw/blob/8ee300b25a0237fc037c90aa24ff1a1067bca8b7/crates/zeroclaw-runtime/src/agent/loop_.rs#L1955), and the first context-window exit in [`reliable.rs`](https://github.com/zeroclaw-labs/zeroclaw/blob/8ee300b25a0237fc037c90aa24ff1a1067bca8b7/crates/zeroclaw-providers/src/reliable.rs#L1320).

### 🔴 Blocking — Keep endpoint grammar inside each locale

`render_provider_failure_lead` constructs `" at {endpoint}"` in English before passing it to every Fluent catalog. Spanish, French, Japanese, and Chinese messages therefore interpolate English grammar. Please pass the endpoint as data and let each locale own the complete sentence or endpoint-bearing variant.

Relevant path: [`i18n.rs`](https://github.com/zeroclaw-labs/zeroclaw/blob/8ee300b25a0237fc037c90aa24ff1a1067bca8b7/crates/zeroclaw-runtime/src/i18n.rs#L367).

### 🟡 Warning — Reconcile the public metadata with the repaired head

Both current commits include a `Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>` trailer, which is not permitted in ZeroClaw public contribution commits. Please remove those trailers, and remove the `(#9001)` review-history breadcrumbs from the new catalog section comments. The PR body should also stop claiming that the CLI and web share the repaired path, that all retry envelopes are removed from user-visible output, and that there is no remaining boundary-test gap until the implementation and evidence support those statements. Its label snapshot should include the live `needs-author-action` label.

**Comment by Audacity88** [comment]
@wm0018, the August 9 review remains unanswered, and GitHub now reports the branch as dirty.

Marking this as a stale candidate. Please address the remaining provider-diagnostics findings, refresh the branch against current master, or reply with your plan by August 28, 2026.

The provider diagnostics issue remains valid. A stale closure would retire this branch, not reject the underlying bug fix.

**Comment by vyahhi** [comment]
[AUTO] Supersedes: PR #10234 explicitly supersedes this provider-diagnostics implementation. Shared objective: replace the generic exhausted-retry envelope with concise cause-specific provider failures. Difference: #10234 preserves typed causes through the reliable layer and projects them through current CLI and RPC/ZeroCode localization boundaries, while this older branch retains unresolved review and conflict state.
